test-app/tests/integration/components/headless-form-test.gts

```diff
@@ -494,6 +514,837 @@ module('Integration Component headless-form', function (hooks) {
     });
   });
 
+  module('validation', function () {
+    interface FormData {
+      firstName?: string;
+      lastName?: string;
+    }
+
+    const validateFormCallbackSync = ({ firstName }: FormData) => {
+      const firstNameErrors = [];
+      if (firstName == undefined) {
+        firstNameErrors.push({
+          type: 'required',
+          value: firstName,
+          message: 'firstName is required!',
+        });
+      } else {
+        if (firstName.charAt(0).toUpperCase() !== firstName.charAt(0)) {
+          firstNameErrors.push({
+            type: 'uppercase',
+            value: firstName,
+            message: 'firstName must be upper case!',
+          });
+        }
+
+        if (firstName.toLowerCase() === 'foo') {
+          firstNameErrors.push({
+            type: 'notFoo',
+            value: firstName,
+            message: 'Foo is an invalid firstName!',
+          });
+        }
+      }
+
+      return firstNameErrors.length > 0
+        ? { firstName: firstNameErrors }
+        : undefined;
+    };
+
+    const validateFormCallbackAsync = async (data: FormData) => {
+      return await validateFormCallbackSync(data);
+    };
+
+    const validateFieldCallbackSync = (
+      value: string | undefined,
+      field: string
+    ) => {
+      const errors = [];
+      if (value == undefined) {
+        errors.push({
+          type: 'required',
+          value,
+          message: `${field} is required!`,
+        });
+      } else {
+        if (value.charAt(0).toUpperCase() !== value.charAt(0)) {
+          errors.push({
+            type: 'uppercase',
+            value,
+            message: `${field} must be upper case!`,
+          });
+        }
+
+        if (value.toLowerCase() === 'foo') {
+          errors.push({
+            type: 'notFoo',
+            value,
+            message: `Foo is an invalid ${field}!`,
+          });
+        }
+      }
+
+      return errors.length > 0 ? errors : undefined;
+    };
+
+    const validateFieldCallbackAsync = async (
+      value: string | undefined,
+      field: string
+    ) => {
+      return await validateFieldCallbackSync(value, field);
+    };
+
+    module('form @validation callback', function () {
+      [
+        { testType: 'sync', validateCallback: validateFormCallbackSync },
+        { testType: 'async', validateCallback: validateFormCallbackAsync },
```

Comment:
for tests using the async callback, how would someone know to show pending state in the UI while they wait for the validation to complete?

Reply:
Good point! 

There is https://github.com/CrowdStrike/ember-headless-form/issues/13, which was mainly about supporting asynchrony of the submit action itself, but I guess we could/should also account for async validation as part of that as well! Will add a note!